config/v1/types_infrastructure.go

```diff
@@ -500,6 +505,58 @@ type AWSResourceTag struct {
 	Value string `json:"value"`
 }
 
+// DNSAndLBConfiguration contains information about API and API-Int Load Balancers
+// and who is responsible for that configuration.
+// +union
+type DNSAndLBConfiguration struct {
+	// provider determines which DNS solution is in use for this cluster.
+	// When the user has pre-created the Load Balancers for API and API-Int and the
+	// the DNS solution has already been configured, this field is set to "UserProvided".
+	// When the cluster is resposible for its own DNS configuration and LB creation,
+	// this field is set to "ClusterProvided".
+	// +default="ClusterProvided"
+	// +kubebuilder:default:="ClusterProvided"
+	// +kubebuilder:validation:Enum="UserProvided";"ClusterProvided"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	Provider DNSAndLBProviderType `json:"provider,omitempty"`
+
+	// userProvided contains information about the user created LoadBalancers.
+	// These values are expected to be set only when provider is set to "UserProvided"
+	UserProvided *LBConfigType `json:"userProvided,omitempty"`
+
+	// clusterProvided contains information about the OpenShift cluster created LoadBalancers.
+	ClusterProvided *LBConfigType `json:"clusterProvided,omitempty"`
+}
+
+// DNSAndLBProviderType defines the source of the DNS and LB configuration.
+type DNSAndLBProviderType string
+
+const (
+	// DNSAndLBUserProvided indicates that the user provides the LB and DNS for API and API-Int.
+	DNSAndLBUserProvided DNSAndLBProviderType = "UserProvided"
```

Comment:
Coupling LB & DNS in this manner is problematic. It excludes a possible valid configuration of user-provided load balancer with installer-provisioned DNS.

Reply:
Updated the type name and the names of the string constants.